internal/k8s/gvrautodetect.go

```diff
@@ -0,0 +1,61 @@
+// Copyright © 2020 VMware
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	projectcontour "github.com/projectcontour/contour/apis/projectcontour/v1"
+
+	ingressroutev1 "github.com/projectcontour/contour/apis/contour/v1beta1"
+
+	serviceapis "sigs.k8s.io/service-apis/api/v1alpha1"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+)
+
+type gvrmap map[schema.GroupVersionResource]cache.SharedIndexInformer
+
+// GVRInformer stores a table of Kubernetes GVR objects to their associated informers.
+type GVRInformer struct {
+	Lookup gvrmap
+}
+
+// DefaultGVRInformer creates a new GVRInformer lookup table and populates with the default
+// GVRs that Contour will try to watch.
+func DefaultGVRInformer(inffactory dynamicinformer.DynamicSharedInformerFactory, serviceAPIs bool) GVRInformer {
+
+	var gvri GVRInformer
```

Comment:
maybe move this closer to its first use around 55 and use a compact literal

gvri := GVRInformer{
   Lookup: make(gvrmap),
}